src/x/cost/limit_manager.go

```diff
@@ -0,0 +1,178 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+package cost
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3/src/cluster/kv"
+	"github.com/m3db/m3/src/cluster/kv/util"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/atomic"
+)
+
+type limitManager struct {
+	sync.RWMutex
+
+	threshold *atomic.Float64
+	enabled   *atomic.Bool
+
+	thresholdWatcher kv.ValueWatch
+	enabledWatcher   kv.ValueWatch
+	closed           bool
+	closeCh          chan struct{}
+	reportInterval   time.Duration
+	metrics          limitManagerMetrics
+}
+
+// NewDynamicLimitManager returns a new LimitWatcher which watches for updates to the cost limit
+// of an operation in KV.
+func NewDynamicLimitManager(
+	store kv.Store,
+	kvLimitKey, kvEnabledKey string,
+	opts LimitManagerOptions,
+) (LimitManager, error) {
+	if opts == nil {
+		opts = NewLimitManagerOptions()
+	}
+	iOpts := opts.InstrumentOptions()
+
+	var (
+		limit            = opts.DefaultLimit()
+		defaultThreshold = float64(limit.Threshold)
+		defaultEnabled   = limit.Enabled
+		m                = &limitManager{
+			threshold:      atomic.NewFloat64(defaultThreshold),
+			enabled:        atomic.NewBool(defaultEnabled),
+			closeCh:        make(chan struct{}),
+			reportInterval: iOpts.ReportInterval(),
+			metrics:        newLimitManagerMetrics(iOpts.MetricsScope()),
+		}
+	)
+
+	watchOpts := util.NewOptions().SetLogger(iOpts.Logger())
+	thresholdWatcher, err := util.WatchAndUpdateAtomicFloat64(
+		store,
+		kvLimitKey,
+		m.threshold,
+		defaultThreshold,
+		watchOpts.SetValidateFn(opts.ValidateLimitFn()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to watch key '%s': %v", kvLimitKey, err)
+	}
+	m.thresholdWatcher = thresholdWatcher
+
+	enabledWatcher, err := util.WatchAndUpdateAtomicBool(
+		store,
+		kvEnabledKey,
+		m.enabled,
+		defaultEnabled,
+		watchOpts,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("unable to watch key '%s': %v", kvEnabledKey, err)
+	}
+	m.enabledWatcher = enabledWatcher
```

Comment:
What's your desired goal state with these? 

```
m.enabledWatcher = enabledWatcher
return m, nil
```
?

Reply:
Btw, bit of a personal rant (not to be taken too seriously :) )

My personal strong opinion about strong style opinions of this variety is that they should be fixed if one of the following is met:

1. Readability is *significantly* impacted (and I would consider that rare--think a crazy long line).
2. They violate a rule imposed by an automated tool (e.g. gofmt).
3. They violate a rule imposed by a standard style guide for the project.

Otherwise, I wouldn't really consider it worth the time.

That said, all of that is just my opinion, and I will adjust these if you like!

Reply:
Probably something like this:

```
   return nil, fmt.Errorf("unable to watch key '%s': %v", kvEnabledKey, err)
  }
 
 m.enabledWatcher = enabledWatcher
 return m, nil
}
```

I think general style approach we were doing is always newline after a `}`, but that may be a query-specific quirk.

Reply:
Haha sounds good. I'll do that for future PR's, but I think I'm going to leave them here, since there are probably more instances than we can reasonably find without an automated tool (due to the fact that this is copied in code).